File: leetcode_problems/typescript/src/lc95_unique_bst_2.ts
function generateTrees(n: number): Array<TreeNode | null> {
  function rec(i: number, n: number): Array<TreeNode | null> {
    if (i > n) {
      return [null];
    }

    let nodes: Array<TreeNode> = [];
    for (let j = i; j <= n; j++) {
      let lefts = rec(i, j - 1);
      let rights = rec(j + 1, n);
      for (let left of lefts) {
        for (let right of rights) {
          nodes.push(new TreeNode(j, left, right));
        }
      }
    }
    return nodes;
  }
  return rec(1, n);
}

File: leetcode_problems/cpp/lc95_unique_bst_2.cpp
#include <vector>
using std::vector;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    vector<TreeNode *> generate(int i, int n)
    {
        if (i > n)
        {
            return {nullptr};
        }

        vector<TreeNode *> nodes;
        for (int j = i; j <= n; j++)
        {
            vector<TreeNode *> lefts = generate(i, j - 1);
            vector<TreeNode *> rigths = generate(j + 1, n);
            for (TreeNode *left : lefts)
            {
                for (TreeNode *right : rigths)
                {
                    TreeNode *head = new TreeNode(j);
                    head->left = left;
                    head->right = right;
                    nodes.push_back(head);
                }
            }
        }
        return nodes;
    }

    vector<TreeNode *> generateTrees(int n)
    {
        return generate(1, n);
    }
};